### ISSUE TITLE
Support multiple platforms

### ISSUE THREAD

**Issue by alabuzhev** [FarGroup/FarManager#3]
A good enough and distant enough goal for the 1st issue here.

**Comment by nightroman** [comment]
It's a good move to enable Issues. Mantis bug tracker is for some "special, involved, audience". Issues are for "normal GitHub account owners" :)

As for this issue, it's a good one and probably doable, after all. FarNet and all its modules including PowerShellFar (with the next x-plat PS v6) are looking forward to this development!

**Comment by michaellukashov** [comment]
just for start: https://github.com/elfmz/far2l
I use it daily on Linux machine

**Comment by alabuzhev** [comment]
Michael, that fork:
* based on the obsolete, unsupported and incompatible 2.0 branch,
* uses questionable (to put it mildly) porting techniques,
* licensed under GPL.

Good to know that it works for you. For the purpose of this issue, however, all that renders it useless.

**Comment by dowsuz** [comment]
What about using more advanced terminal keys ?

http://www.leonerd.org.uk/code/libtermkey/

http://bazaar.leonerd.org.uk/c/libtermkey/

https://github.com/o11c/libtermkey

https://github.com/xinhaoyuan/libtermkey

**Comment by techtonik** [comment]
> What about using more advanced terminal keys ?

Parsing ANSI sequences for key presses is a dead way - it is slow, unreliable, and not compatible with shortcuts for different keyboard layouts. I'd aim for better interface for keyboard event input and insert terminal compatibility clutches later.

Missing Far for a long time, and knowing that it is being a marginal product, I found https://i3wm.org/ a best alternative to my mouseless workflow. If Far could just merge in this environment - that would be awesome to use a code editor and quick interface to move files around a project.

So, a way to get reliable 120 FPS in Linux console with no input lags would be to either try i3wm or research Wayland and propose changes there while it is not too late. Because guys are obsessed with graphics and Vulkan doesn't mean that console experience should suffer. And now with Far expertise it is a best way to rethink how it should be done.

**Comment by alexmyczko** [comment]
maybe one can learn, if someone remembers opencubicplayer for dos/windows, it was successfully ported to linux/*nix: https://github.com/mywave82/opencubicplayer

**Comment by ikaman** [comment]
yeah i wish we had this as well. far2l and mc on linux are far inferior in my mind

**Comment by yuchdev** [comment]
Let me know if you decide to start porting Far. I'm using it since 2002 and was always dreaming to replace the clunky MC on Linux. You can count on my contribution.
I can help with porting the build system on some platform-independent toolchain, like CMake or the new one, Meson.

**Comment by unxed** [comment]
> * based on the obsolete, unsupported and incompatible 2.0 branch,
> * uses questionable (to put it mildly) porting techniques,
> * licensed under GPL.

For reference. Previously, the ecosystem lacked the capabilities for a full-fledged port with a full UX. Now there's [win32-input-mode](https://github.com/microsoft/terminal/blob/main/doc/specs/%234999%20-%20Improved%20keyboard%20handling%20in%20Conpty.md), the [kitty keyboard protocol](https://sw.kovidgoyal.net/kitty/keyboard-protocol/), [bracketed-paste](https://en.wikipedia.org/wiki/Bracketed-paste) for pasting from and [osc52](https://github.com/microsoft/terminal/issues/2946) for copying to clipboard, there's even the [kitty graphics protocol](https://sw.kovidgoyal.net/kitty/graphics-protocol/) for implementing Picture Viewer and [sixel graphics](https://en.wikipedia.org/wiki/Sixel) for [graceful degradation](https://github.com/hpjansson/chafa/issues/312). From a purely technical standpoint, there's no fundamental reason why a 3.0 version port can't be made now using clean and accurate porting techniques and without any GPL'ed code.

In far2l we support all specs mentioned above. Both in tty backend and in internal virtual terminal.

> not compatible with shortcuts for different keyboard layouts

Actually, both win32-input-mode and kitty keyboard protocol ARE compatible. You need to use the VirtualKeyCode field in win32-input-mode and base layout key field in the kitty protocol, and for non-Latin hotkeys use xlat tables. And we in far2l team have no problems with speed and reliability for years.

**Comment by unxed** [comment]
> not compatible with shortcuts for different keyboard layouts

Actually, both win32-input-mode and kitty keyboard protocol ARE compatible. You need to use the VirtualKeyCode field in win32-input-mode and base layout key field in the kitty protocol, and for non-Latin hotkeys use xlat tables. And we in far2l team have no problems with speed and reliability for years.

**Comment by unxed** [comment]
Also I would like to share my vision of the future of the ecosystem.

At the moment there are three projects inherited from the Far 2 sources:

- Far 3
- [far2l](https://github.com/elfmz/far2l/) by @elfmz
- [far2m](https://github.com/shmuz/far2m) by @shmuz , fork of far2l

As far as I understand, the source codes of the first and the last are slowly getting closer, and that's great.

But the codebases of Far 3 and far2l are unlikely to behave the same way, for reasons that are more political than technical. Far 3 is about the Cathedral, while the Bazaar part of the ecosystem has been moved into scripts and plugins. In far2l the patch acceptance policy is more permissive, and a larger portion of the functionality is implemented in C/C++.

Therefore it seems more reasonable to negotiate standards (and possibly plugin APIs in the future). For example, full UX compatibility inside each other's consoles looks like a nice intermediate goal. I also proposed [some improvements](https://github.com/elfmz/far2l/issues/3137) to ConEmu that would make the far2l UX perfect (and we already know how to access the clipboard under WSL using vbs script).
